### ISSUE TITLE
Implementation of struct filter::generic::FreeRng needed?

### ISSUE THREAD

**Issue by steindev** [ComputationalRadiationPhysics/picongpu#4439]
While working with Rng based particle filters in `ManipulateDerive<...>` in `speciesInitialization.param` I noted, that the implementation of filter::generic::FreeRng is broken.

[The shorthand definition of RngType in line 89 of include/picongpu/particles/filter/generic/FreeRng.hpp](https://github.com/ComputationalRadiationPhysics/picongpu/blob/59e9b53605f9a5c1bf271eeb055bc74370a99052/include/picongpu/particles/filter/generic/FreeRng.hpp#L89) is failing.

However, it is not needed anyways, as `RngType` can be replaced with `auto` and `ALPAKA_DECAY_T(decltype(rng))` in subsequent lines, just as in `include/picongpu/particles/manipulators/generic/FreeRng.hpp`

The point of this issue is to discuss whether the implementation of `filter::generic::FreeRng` is needed at all, as it is identical to the implementation of `manipulators::generic::FreeRng`, except for the return type of `operator()` in `struct generic::acc::FreeRng : private T_Functor`.

**Comment by psychocoderHPC** [comment]
As you mentioned the return type is different. Filters and Manipulator have therefore a different interface. PIConGPU is compile time validating if the interface is fitting https://github.com/ComputationalRadiationPhysics/picongpu/blob/59e9b53605f9a5c1bf271eeb055bc74370a99052/include/picongpu/particles/filter/IUnary.def#L40

Could you please post the compile time error!

**Comment by psychocoderHPC** [comment]
error message is:

```
picongpu/include/picongpu/../picongpu/particles/filter/generic/FreeRng.hpp(89): error: class "picongpu::particles::functor::misc::Rng<picongpu::particles::atomicPhysics::DistributionFloat>" has no member "RandomGen"
          detected during:
            instantiation of class "picongpu::particles::filter::generic::FreeRng<T_Functor, T_Distribution> [with T_Functor=picongpu::particles::filter::FunctorEachSecondParticle, T_Distribution=pmacc::random::distributions::Uniform<picongpu::float_X, pmacc::random::methods::RngPlaceholder>]" 
/opt/spack-modules/linux-ubuntu20.04-skylake_avx512/gcc-10.2.0/boost-1.75.0-rhit3f5bvbmaffhbwi27c6vhvf5ip7iw/include/boost/mpl/aux_/preprocessed/gcc/apply_wrap.hpp(38): here
            instantiation of class "boost::mpl::apply_wrap1<F, T1> [with F=picongpu::particles::filter::generic::FreeRng<picongpu::particles::filter::FunctorEachSecondParticle, pmacc::random::distributions::Uniform<picongpu::float_X, pmacc::random::methods::RngPlaceholder>>, T1=pmacc::particles::meta::FindByNameOrType_t<picongpu::VectorAllSpecies, picongpu::PIC_Electrons, pmacc::errorHandlerPolicies::ThrowValueNotFound>]" 
/opt/spack-modules/linux-ubuntu20.04-skylake_avx512/gcc-10.2.0/boost-1.75.0-rhit3f5bvbmaffhbwi27c6vhvf5ip7iw/include/boost/mpl/aux_/preprocessed/gcc/apply.hpp(44): here
```

For unknown reason `atomicPhysics::DistributionFloat` is kicking in and producing the compile error.

**Comment by psychocoderHPC** [comment]
@steindev suggestion for the fix is correct, I missed updating the filters in picongpu during the refactoring of the RNG generator in the past.

**Comment by psychocoderHPC** [comment]
@steindev, In general, you are right the implementations of both files rng filter and manipulator are nearly equal but the behavior of both are different.
At all the classes RNGFree, FreeTotalCell,... functors and filters are workarounds to solve for now a problem that we can currently not transfer special knowledge like the state of the random number generator into user-defined functors.
I have already ideas in my head about how to solve this fully, we do not have an issue with this because I do not know how to describe the solution in the text.
We have the same issue in the collision where we build our own context element to provide the random number generator and global offsets into the functors.

**Comment by PrometheusPi** [comment]
solved by #4440

**Comment by steindev** [comment]
> @steindev, In general, you are right the implementations of both files rng filter and manipulator are nearly equal but the behavior of both are different.

Well I must have had tomatoes on my eyes. You are right, there are a couple more differences I did not appreciate but are important.
